python/dlisio/ext/lis.cpp

```diff
@@ -120,73 +156,252 @@ noexcept (false) {
+noexcept (false) {
+    std::array< unsigned char, sizeof(T) > buf;
+    auto* dst = static_cast< unsigned char* >(buf.data());
 
-        else {
-            int src_skip, dst_skip;
-            const char localfmt[] = {*f, '\0'};
-            lis_packflen(localfmt, ptr, &src_skip, &dst_skip);
-            assert_overflow(ptr, end, src_skip);
-            lis_packf(localfmt, ptr, dst);
-            dst += dst_skip;
-            ptr += src_skip;
-            ++f;
+    const char* indexfmt = fconf.indexfmt.c_str();
+    read_sample( indexfmt, ptr, end, dst );
+
+    T idx;
+    std::memcpy(&idx, buf.data(), sizeof(T));
+    return static_cast< double >(idx);
+}
+
+const char* read_index( const char*        src,
+                        const char*        end,
+                        const frameconfig& fconf,
+                        indexchannel&      channel )
+noexcept (false) {
+    using namespace std;
+
+    double x;
+    const char* indexfmt = fconf.indexfmt.c_str();
+    switch (*indexfmt) {
+        case LIS_FMT_I8:     { x = from_buf< int8_t  >(src, end, fconf); break; }
+        case LIS_FMT_I16:    { x = from_buf< int16_t >(src, end, fconf); break; }
+        case LIS_FMT_I32:    { x = from_buf< int32_t >(src, end, fconf); break; }
+        case LIS_FMT_F16:    { x = from_buf< float   >(src, end, fconf); break; }
+        case LIS_FMT_F32:    { x = from_buf< float   >(src, end, fconf); break; }
+        case LIS_FMT_F32LOW: { x = from_buf< float   >(src, end, fconf); break; }
+        case LIS_FMT_F32FIX: { x = from_buf< float   >(src, end, fconf); break; }
+        case LIS_FMT_BYTE:   { x = from_buf< uint8_t >(src, end, fconf); break; }
+        case LIS_FMT_STRING:
+        case LIS_FMT_MASK: {
+            const auto msg = "Unsupported datatype for index channel, was: ";
+            throw std::runtime_error(msg + fconf.indexfmt);
+        }
+        default: {
+            const auto msg = "Invalid datatype for index channel, was: ";
+            throw std::runtime_error(msg + fconf.indexfmt);
         }
     }
+    channel.update(x);
+    return src;
 }
 
-void read_data_record( const std::string& fmt,
-                       const lis::record& src,
+void read_frame( const char*&       ptr,
+                 const char*        end,
+                 unsigned char*&    dst,
+                 indexchannel&      index,
+                 const frameconfig& fconf )
+noexcept (false) {
+    bool fast = ( fconf.samples > 1 ) ? true : false;
+
+    /* Writing the index to dst
+     *
+     * The trivial case is when the index channel is recorded as a regular
+     * channel (mode=0) and we don't have fast channels. Then, there is no need
+     * for the intermediate storage of the index values and we can write
+     * directly from record (ptr) to dst.
+     */
+    if ( fconf.mode == 0 and not fast ) {
+        const char* indexfmt = fconf.indexfmt.c_str();
+        read_sample( indexfmt, ptr, end, dst );
```

Comment:
Nope! I'm concerned about the performance penalty of always writing to the intermediate storage, and then to the to dst. It's one extra copy (+ boilerplate) for every single value. It makes no sense to pay that penalty when we don't need too. Especially since mode=0 + samples=1 is the most common configuration